### PR TITLE
Mesh shaders

### PR DESCRIPTION
**Pull request by Firestar99**
# Depends on https://github.com/Rust-GPU/rust-gpu/pull/43

Diff without dependent PR: https://github.com/Firestar99/rust-gpu-new/compare/update-rspirv...Firestar99:rust-gpu-new:mesh-shaders

changelog:
```
- [PR#44](https://github.com/Rust-GPU/rust-gpu/pull/44) added support for mesh and task shaders
```

### CODE REVIEW

**Comment by LegNeato** [comment]
It would be good to have a test, changelog entry, and an example for this. Might be good to do a blog post too!

**Comment by Firestar99** [comment]
Plenty of tests already exist in this PR and I added a changelog entry in the description. I can make you a quick example shader later.

**Comment by Firestar99** [comment]
There are two vendor-specific issues I know of with regards to "payloads" (sending data from task to mesh shader):
* AMD on Windows assumes the payload is always a struct, even if the spec says it could be anything like an uint. It not being a struct will segfault the driver on pipeline creation. If you define a struct with just a single primitive member, like a uint, rust-gpu will "inline" the struct and make the payload be a uint. Workaround: add a second dummy member.
* RenderDoc suffers from the same issue as above, and will display something like "shader likely compiled by some old weird dxc version"
* RenderDoc assumes the struct for the payload is never constructed, only it's fields written to. For it's debugging, it will inject a new field into said struct to emit some debug info. If you just assign fields to that struct it's fine, but it won't fixup the `OpConstruct` statements to make a new struct, resulting in invalid spirv and RenderDoc complaining. Rust-gpu merges structs with the same members, making it rather easy to trip onto this one. 

Suggested fix: Wrap all task shader payloads in a separate struct that is not optimized away.

**Comment by LegNeato** [comment]
> Plenty of tests already exist in this PR and I added a changelog entry in the description. I can make you a quick example shader later.

Oh weird, they didn't show up on mobile!

File: crates/rustc_codegen_spirv/src/spirv_type_constraints.rs [hunk 1]
```diff
@@ -945,6 +953,182 @@ pub fn instruction_signatures(op: Op) -> Option<&'static [InstSig<'static>]> {
         | Op::SubgroupAvcSicGetInterRawSadsINTEL => {
             reserved!(SPV_INTEL_device_side_avc_motion_estimation);
         }
+        // SPV_EXT_mesh_shader
+        Op::EmitMeshTasksEXT | Op::SetMeshOutputsEXT => {
+            // NOTE(eddyb) we actually use these despite not being in the standard yet.
+            // reserved!(SPV_EXT_mesh_shader)
```
Comment (LegNeato) [inline crates/rustc_codegen_spirv/src/spirv_type_constraints.rs]: Should this be uncommented?

Comment (Firestar99) [reply crates/rustc_codegen_spirv/src/spirv_type_constraints.rs]: Removed the comment and code. Previously, that was `reserved!()` aka. not implemented and now it should just do nothing when validating type constraints on these intrinsics, as they're already validated by the functions representing these intrinsics (and so do most other intrinsics as well).

Comment (Firestar99) [reply crates/rustc_codegen_spirv/src/spirv_type_constraints.rs]: And rebased.

**Comment by LegNeato** [comment]
@eddyb or @schell , I don't know anything about mesh shaders...can either of you review?

**Comment by Firestar99** [comment]
I think nobody but me knows about mesh shaders...
I wouldn't worry too much about correctness of the mesh shader stuff, I've tested it plenty so hopefully all issues have already been caught.

**Comment by LegNeato** [comment]
Let's land this and iterate if needed, it has been sitting too long.

**Review by LegNeato — APPROVED** [review]
:shipit:

**Pull request merged**